src/util/DeviceFingerprint.js

```diff
@@ -0,0 +1,79 @@
+/*!
+ * Copyright (c) 2015-2016, Okta, Inc. and/or its affiliates. All rights reserved.
+ * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
+ *
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+/*jshint latedef:false */
+/*global JSON */
+
+define(['vendor/lib/q', 'jquery'], function (Q, $) {
+
+  return {
+    generateDeviceFingerprint: function (oktaDomainUrl) {
+      var deferred = Q.defer();
+
+      // Create iframe
+      var $iframe = $('<iframe>', {
+        id: 'okta_fingerprint_iframe',
+        style: 'display: none;'
+      });
+      $iframe.appendTo('.auth-content');
+
+      function removeIframe() {
+        $iframe.off();
+        $iframe.remove();
+        window.removeEventListener('message', onMessageReceivedFromOkta, false);
+      }
+
+      function handleError(reason) {
+        removeIframe();
+        deferred.reject(reason);
+      }
+
+      function onMessageReceivedFromOkta(event) {
+        /*jshint maxcomplexity:7*/
+        if (!event || !event.data || event.origin != oktaDomainUrl) {
+          handleError('no data');
+        }
+        try {
+          var message = JSON.parse(event.data);
+          if (message && message.type == 'FingerprintServiceReady') {
+            var actionMessage = {};
+            actionMessage.type = 'GetFingerprint';
+            sendMessageToOkta(actionMessage);
+          } else if (message && message.type == 'FingerprintAvailable') {
+            removeIframe();
+            deferred.resolve(message.fingerprint);
+          } else {
+            handleError('no data');
+          }
+        } catch (error) {
+          handleError(error);
+          //Ignore any errors since we could get other messages too
```

Comment:
Hmmm, are we ignoring it though? It seems like if we receive other types of messages, or get an error, we'll immediately reject and not process anything else right?